Parsers/ASimWebSession/Parsers/vimWebSessionSonicWallFirewall.yaml

```diff
@@ -0,0 +1,450 @@
+  This ASIM parser supports filtering and normalizing SonicWall SonicOS ArcSight-formatted syslog ingested by Microsoft Sentinel to the ASIM WebSession normalized schema.
+ParserName: vimWebSessionSonicWallFirewall
+EquivalentBuiltInParser: _Im_WebSession_SonicWallFirewall
+ParserParams:
+  - Name: starttime
+    Type: datetime
+    Default: datetime(null)
+  - Name: endtime
+    Type: datetime
+    Default: datetime(null)
+  - Name: srcipaddr_has_any_prefix
+    Type: dynamic
+    Default: dynamic([])
+  - Name: ipaddr_has_any_prefix
+    Type: dynamic
+    Default: dynamic([])
+  - Name: url_has_any
+    Type: dynamic
+    Default: dynamic([])
+  - Name: httpuseragent_has_any
+    Type: dynamic 
+    Default: dynamic([])
+  - Name: eventresultdetails_in
+    Type: dynamic
+    Default: dynamic([])
+  - Name: eventresult
+    Type: string
+    Default: '*'
+  - Name: disabled
+    Type: bool
+    Default: false
+ParserQuery: |
+  let parser=(
+    starttime:datetime=datetime(null),
+    endtime:datetime=datetime(null),
+    srcipaddr_has_any_prefix:dynamic=dynamic([]),
+    ipaddr_has_any_prefix:dynamic=dynamic([]), 
+    url_has_any:dynamic=dynamic([]),
+    httpuseragent_has_any:dynamic=dynamic([]),
+    eventresultdetails_in:dynamic=dynamic([]),
+    eventresult:string='*',
+    disabled:bool=false
+    )
+    {
+      let src_or_any=set_union(srcipaddr_has_any_prefix, ipaddr_has_any_prefix);
+      let Actions=datatable(fw_action:string, DvcAction:string, EventSeverity:string)
+        [ "\"forward\"", "Allow", "Informational"
+        , "\"mgmt\"", "Other", "Informational"
+        , "\"NA\"", "Other", "Informational"
+        , "\"drop\"", "Drop", "Low"
+      ];
+      CommonSecurityLog
+      | where not(disabled)
+      | where DeviceVendor == "SonicWall"
+      | where DeviceEventClassID in (14, 97)
+      | where (isnull(starttime) or TimeGenerated >= starttime) and (isnull(endtime) or TimeGenerated <= endtime)
+      | where (isnotempty(SourceIP) or isnotempty(DestinationIP))
+      | extend temp_SrcMatch = has_any_ipv4_prefix(SourceIP, src_or_any)
+        , temp_DstMatch = has_any_ipv4_prefix(DestinationIP, ipaddr_has_any_prefix)
+      | extend ASimMatchingIpAddr = case(array_length(src_or_any) == 0, "-",
+                                        temp_SrcMatch and temp_DstMatch, "Both",
+                                        temp_SrcMatch, "SrcIpAddr",
+                                        temp_DstMatch, "DstIpAddr",
+                                        "No match")
+      | where ASimMatchingIpAddr != "No match"
+      | project-away temp_*
+      | extend EventResultDetails = "NA"
+      | where (array_length(eventresultdetails_in) == 0 or tostring(EventResultDetails) has_any(eventresultdetails_in))
+      | where (array_length(httpuseragent_has_any) == 0 or RequestClientApplication has_any (httpuseragent_has_any))
+      | where Protocol has_any(dynamic(["udp/http", "tcp/http", "udp/https", "tcp/https"]))

```

Comment:
can we move all these where conditions to earlier 'where' filter and separate all of the with 'and'?